resolver/custom_dns_resolver.go

```diff
@@ -85,12 +92,22 @@ func (r *CustomDNSResolver) processRequest(request *model.Request) *model.Respon
 	domain := util.ExtractDomain(question)
 
 	for len(domain) > 0 {
```

Comment:
Let's add a context cancellation check here. That way if we're in an infinite CNAME loop, we'll have an exit strategy:
```suggestion
	for len(domain) > 0 {
		if err := ctx.Err(); err != nil {
			return nil, err
		}

```

We also need another change to actually trigger it, since ATM we never cancel the context.  
Can you also add a timeout in `Server.resolve`, just above the `switch`?
```go
	ctx, cancel := context.WithTimeout(ctx, 100*s.cfg.Upstreams.Timeout.ToDuration())
	defer cancel()

	switch {
		// ...
```

That way we'll reduce the impact of a bad config, and generally having a high global timeout per request is a good safety measure.